File: JavaScript/Configuration/API_Config.js
// API_Config.js
import axios from 'axios';
import NodeCache from 'node-cache';
import logger from './logger.js';
import Decimal from 'decimal.js';
import fs from 'fs/promises';

class API_Config {
    constructor(configuration = null) {
        this.configuration = configuration; // Configuration settings
        this.session = axios.create();
        this.priceCache = new NodeCache({ stdTTL: 300, checkperiod: 60 }); // Real-time prices cache
        this.tokenSymbolCache = new NodeCache({ stdTTL: 86400, checkperiod: 3600 }); // Token symbols cache

        // API configuration detailing different API providers and their settings
        this.apiConfigs = {
            "binance": {
                "base_url": "https://api.binance.com/api/v3",
                "success_rate": 1.0,
                "weight": 1.0,
                "rate_limit": 1200, // Max requests per minute
            },
            "coingecko": {
                "base_url": "https://api.coingecko.com/api/v3",
                "api_key": this.configuration.COINGECKO_API_KEY,
                "success_rate": 1.0,
                "weight": 0.8,
                "rate_limit": 50, // Max requests per minute
            },
            "coinmarketcap": {
                "base_url": "https://pro-api.coinmarketcap.com/v1",
                "api_key": this.configuration.COINMARKETCAP_API_KEY,
                "success_rate": 1.0,
                "weight": 0.7,
                "rate_limit": 333, // Max requests per minute
            },
            "cryptocompare": {
                "base_url": "https://min-api.cryptocompare.com/data",
                "api_key": this.configuration.CRYPTOCOMPARE_API_KEY,
                "success_rate": 1.0,
                "weight": 0.6,
                "rate_limit": 80, // Max requests per minute
            },
        };

        // Initialize rate limiters for each API provider based on their rate limits
        this.rateLimiters = {};
        for (let provider in this.apiConfigs) {
            const config = this.apiConfigs[provider];
            this.rateLimiters[provider] = {
                tokens: config.rate_limit,
                lastRefill: Date.now(),
                tokensPerInterval: config.rate_limit,
                interval: 60000, // 60 seconds
            };
        }
    }

    async getTokenSymbol(web3, tokenAddress) {
        if (this.tokenSymbolCache.has(tokenAddress)) {
            return this.tokenSymbolCache.get(tokenAddress);
        }

        if (this.configuration.TOKEN_SYMBOLS[tokenAddress]) {
            const symbol = this.configuration.TOKEN_SYMBOLS[tokenAddress];
            this.tokenSymbolCache.set(tokenAddress, symbol);
            return symbol;
        }

        try {
            const abiPath = this.configuration.ERC20_ABI;
            const content = await fs.readFile(abiPath, 'utf-8');
            const abi = JSON.parse(content);
            const contract = new web3.eth.Contract(abi, tokenAddress);
            const symbol = await contract.methods.symbol().call();
            this.tokenSymbolCache.set(tokenAddress, symbol);
            return symbol;
        } catch (e) {
            logger.error(`Error getting symbol for token ${tokenAddress}: ${e.message}`);
            return null;
        }
    }

    async getRealTimePrice(token, vsCurrency = "eth") {
        const cacheKey = `price_${token}_${vsCurrency}`;
        if (this.priceCache.has(cacheKey)) {
            return this.priceCache.get(cacheKey);
        }

        let prices = [];
        let weights = [];

        for (let source in this.apiConfigs) {
            try {
                const price = await this._fetchPrice(source, token, vsCurrency);
                if (price) {
                    const config = this.apiConfigs[source];
                    prices.push(price);
                    weights.push(config.weight * config.success_rate);
                }
            } catch (e) {
                logger.error(`Error fetching price from ${source}: ${e.message}`);
                this.apiConfigs[source].success_rate *= 0.9; // Penalize failed source
            }
        }

        if (prices.length === 0) {
            logger.warn(`No valid prices found for ${token}!`);
            return null;
        }

        // Calculate weighted average price
        const weightedSum = prices.reduce((acc, p, idx) => acc + p * weights[idx], 0);
        const totalWeight = weights.reduce((acc, w) => acc + w, 0);
        const weightedPrice = weightedSum / totalWeight;

        this.priceCache.set(cacheKey, new Decimal(weightedPrice));
        return this.priceCache.get(cacheKey);
    }

    async _fetchPrice(source, token, vsCurrency) {
        const config = this.apiConfigs[source];
        if (!config) {
            logger.debug(`API configuration for ${source} not found.`);
            return null;
        }

        let url = '';
        let params = {};
        let headers = {};

        if (source === "coingecko") {
            url = `${config.base_url}/simple/price`;
            params = { ids: token, vs_currencies: vsCurrency };
        } else if (source === "coinmarketcap") {
            url = `${config.base_url}/cryptocurrency/quotes/latest`;
            params = { symbol: token.toUpperCase(), convert: vsCurrency.toUpperCase() };
            headers = { 'X-CMC_PRO_API_KEY': config.api_key };
        } else if (source === "cryptocompare") {
            url = `${config.base_url}/price`;
            params = { fsym: token.toUpperCase(), tsyms: vsCurrency.toUpperCase(), api_key: config.api_key };
        } else if (source === "binance") {
            url = `${config.base_url}/ticker/price`;
            const symbol = `${token.toUpperCase()}${vsCurrency.toUpperCase()}`;
            params = { symbol: symbol };
        } else {
            logger.warn(`Unsupported price source: ${source}`);
            return null;
        }

        const response = await this.makeRequest(source, url, { params, headers });
        if (source === "coingecko") {
            return new Decimal(response[token][vsCurrency]);
        } else if (source === "coinmarketcap") {
            const data = response.data[token.toUpperCase()].quote[vsCurrency.toUpperCase()].price;
            return new Decimal(data);
        } else if (source === "cryptocompare") {
            return new Decimal(response[vsCurrency.toUpperCase()]);
        } else if (source === "binance") {
            return new Decimal(response.price);
        }

        return null;
    }

    async makeRequest(providerName, url, { params = {}, headers = {} } = {}, maxAttempts = 5, backoffFactor = 1.5) {
        await this._acquireRateLimit(providerName);

        for (let attempt = 0; attempt < maxAttempts; attempt++) {
            try {
                const response = await this.session.get(url, { params, headers, timeout: 10000 * (attempt + 1) });
                if (response.status === 429) {
                    const waitTime = Math.pow(backoffFactor, attempt) * 1000;
                    logger.warn(`Rate limit exceeded for ${providerName}, retrying in ${waitTime}ms...`);
                    await new Promise(res => setTimeout(res, waitTime));
                    continue;
                }
                response.status !== 200 && response.raiseForStatus();
                return response.data;
            } catch (e) {
                if (attempt === maxAttempts - 1) {
                    logger.error(`Request failed after ${maxAttempts} attempts: ${e.message}`);
                    throw e;
                }
                const waitTime = Math.pow(backoffFactor, attempt) * 1000;
                logger.warn(`Request attempt ${attempt + 1} failed: ${e.message}. Retrying in ${waitTime}ms...`);
                await new Promise(res => setTimeout(res, waitTime));
            }
        }
    }

    async _acquireRateLimit(providerName) {
        const limiter = this.rateLimiters[providerName];
        if (!limiter) return;

        const now = Date.now();
        if (now - limiter.lastRefill > limiter.interval) {
            limiter.tokens = limiter.tokensPerInterval;
            limiter.lastRefill = now;
        }

        if (limiter.tokens > 0) {
            limiter.tokens -= 1;
            return;
        }

        const waitTime = limiter.interval - (now - limiter.lastRefill);
        logger.warn(`Rate limit reached for ${providerName}. Waiting for ${waitTime}ms...`);
        await new Promise(res => setTimeout(res, waitTime));
        limiter.tokens = limiter.tokensPerInterval - 1;
        limiter.lastRefill = Date.now();
    }

    async fetchHistoricalPrices(token, days = 30) {
        const cacheKey = `historical_prices_${token}_${days}`;
        if (this.priceCache.has(cacheKey)) {
            logger.debug(`Returning cached historical prices for ${token}.`);
            return this.priceCache.get(cacheKey);
        }

        const prices = await this._fetchFromServices(
            (service) => this._fetchHistoricalPrices(service, token, days),
            `historical prices for ${token}`
        );

        if (prices) {
            this.priceCache.set(cacheKey, prices);
        }

        return prices || [];
    }

    async _fetchHistoricalPrices(source, token, days) {
        const config = this.apiConfigs[source];
        if (!config) {
            logger.debug(`API configuration for ${source} not found.`);
            return null;
        }

        if (source === "coingecko") {
            const url = `${config.base_url}/coins/${token}/market_chart`;
            const params = { vs_currency: 'usd', days: days };
            const response = await this.makeRequest(source, url, { params });
            return response.prices.map(price => price[1]);
        } else {
            logger.debug(`Unsupported historical price source: ${source}`);
            return null;
        }
    }

    async getTokenVolume(token) {
        const cacheKey = `token_volume_${token}`;
        if (this.priceCache.has(cacheKey)) {
            logger.debug(`Returning cached trading volume for ${token}.`);
            return this.priceCache.get(cacheKey);
        }

        const volume = await this._fetchFromServices(
            (service) => this._fetchTokenVolume(service, token),
            `trading volume for ${token}`
        );

        if (volume !== null) {
            this.priceCache.set(cacheKey, volume);
        }

        return volume || 0.0;
    }

    async _fetchTokenVolume(source, token) {
        const config = this.apiConfigs[source];
        if (!config) {
            logger.debug(`API configuration for ${source} not found.`);
            return null;
        }

        if (source === "coingecko") {
            const url = `${config.base_url}/coins/markets`;
            const params = { vs_currency: 'usd', ids: token };
            const response = await this.makeRequest(source, url, { params });
            return response.length > 0 ? response[0].total_volume : null;
        } else {
            logger.debug(`Unsupported volume source: ${source}`);
            return null;
        }
    }

    async _fetchFromServices(fetchFunc, description) {
        for (let service in this.apiConfigs) {
            try {
                logger.debug(`Fetching ${description} using ${service}...`);
                const result = await fetchFunc(service);
                if (result) {
                    return result;
                }
            } catch (e) {
                logger.warn(`Failed to fetch ${description} using ${service}: ${e.message}`);
            }
        }
        logger.warn(`Failed to fetch ${description}.`);
        return null;
    }

    async _loadAbi(abiPath) {
        try {
            const content = await fs.readFile(abiPath, 'utf-8');
            const abi = JSON.parse(content);
            logger.debug(`Loaded ABI from ${abiPath} successfully.`);
            return abi;
        } catch (e) {
            logger.error(`Failed to load ABI from ${abiPath}: ${e.message}`);
            throw e;
        }
    }

    async close() {
        try {
            await this.session.close();
        } catch (e) {
            logger.error(`Error closing API session: ${e.message}`);
        }
    }
}

export default API_Config;

File: python/configuration/api_config.py

import asyncio
import json
import aiofiles
import aiohttp
from decimal import Decimal
from cachetools import TTLCache
from typing import Any, Dict, List, Optional
from web3 import AsyncWeb3
import logging

from configuration.configuration import Configuration

logger = logging.getLogger(__name__)


class API_Config:
    def __init__(self, configuration: Optional[Configuration] = None):
        self.configuration = configuration
        self.session = None 
        self.price_cache = TTLCache(maxsize=1000, ttl=300)  # Cache for 5 minutes
        self.token_symbol_cache = TTLCache(maxsize=1000, ttl=86400)  # Cache for 1 day

    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()
            logger.debug("APIconfig session closed.")

        # API configuration
        self.api_configs = {
            "binance": {
                "base_url": "https://api.binance.com/api/v3",
                "success_rate": 1.0,
                "weight": 1.0,
                "rate_limit": 1200,  # Max requests per minute
            },
            "coingecko": {
                "base_url": "https://api.coingecko.com/api/v3",
                "api_key": self.configuration.COINGECKO_API_KEY,
                "success_rate": 1.0,
                "weight": 0.8,
                "rate_limit": 50,  # Max requests per minute
            },
            "coinmarketcap": {
                "base_url": "https://pro-api.coinmarketcap.com/v1",
                "api_key": self.configuration.COINMARKETCAP_API_KEY,
                "success_rate": 1.0,
                "weight": 0.7,
                "rate_limit": 333,  # Max requests per minute
            },
            "cryptocompare": {
                "base_url": "https://min-api.cryptocompare.com/data",
                "api_key": self.configuration.CRYPTOCOMPARE_API_KEY,
                "success_rate": 1.0,
                "weight": 0.6,
                "rate_limit": 80,  # Max requests per minute
            },
        }

        self.api_lock = asyncio.Lock()
        self.rate_limiters = {
            provider: asyncio.Semaphore(config.get("rate_limit", 10))
            for provider, config in self.api_configs.items()
        }

    async def get_token_symbol(self, web3: AsyncWeb3, token_address: str) -> Optional[str]:
        """Get the token symbol for a given token address."""
        if token_address in self.token_symbol_cache:
            return self.token_symbol_cache[token_address]
        if token_address in self.configuration.TOKEN_SYMBOLS:
            symbol = self.configuration.TOKEN_SYMBOLS[token_address]
            self.token_symbol_cache[token_address] = symbol
            return symbol
        try:
            erc20_abi = await self._load_abi(self.configuration.ERC20_ABI)
            contract = web3.eth.contract(address=token_address, abi=erc20_abi)
            symbol = await contract.functions.symbol().call()
            self.token_symbol_cache[token_address] = symbol
            return symbol
        except Exception as e:
            logger.error(f"Error getting symbol for token {token_address}: {e}")
            return None

    async def get_real_time_price(self, token: str, vs_currency: str = "eth") -> Optional[Decimal]:
        """Get real-time price using weighted average from multiple sources."""
        cache_key = f"price_{token}_{vs_currency}"
        if cache_key in self.price_cache:
            return self.price_cache[cache_key]
        prices = []
        weights = []
        async with self.api_lock:
            for source, config in self.api_configs.items():
                try:
                    price = await self._fetch_price(source, token, vs_currency)
                    if price:
                        prices.append(price)
                        weights.append(config["weight"] * config["success_rate"])
                except Exception as e:
                    logger.error(f"Error fetching price from {source}: {e}")
                    config["success_rate"] *= 0.9
        if not prices:
            logger.warning(f"No valid prices found for {token}!")
            return None
        weighted_price = sum(p * w for p, w in zip(prices, weights)) / sum(weights)
        self.price_cache[cache_key] = Decimal(str(weighted_price))
        return self.price_cache[cache_key]

    async def _fetch_price(self, source: str, token: str, vs_currency: str) -> Optional[Decimal]:
        """Fetch the price of a token from a specified source."""
        config = self.api_configs.get(source)
        if not config:
            logger.error(f"API source {source} not configured.")
            return None

        try:
            async with self.session.get(config["base_url"] + f"/simple/price?ids={token}&vs_currencies={vs_currency}") as response:
                if response.status == 200:
                    data = await response.json()
                    price = Decimal(str(data[token][vs_currency]))
                    logger.debug(f"Fetched price from {source}: {price}")
                    return price
                else:
                    logger.error(f"Failed to fetch price from {source}: Status {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Exception fetching price from {source}: {e}")
            return None

    async def make_request(
        self,
        provider_name: str,
        url: str,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        max_attempts: int = 5,
        backoff_factor: float = 1.5,
    ) -> Any:
        """Make HTTP request with exponential backoff and rate limit per provider."""
        rate_limiter = self.rate_limiters.get(provider_name)
        
        if rate_limiter is None:
            logger.error(f"No rate limiter for provider {provider_name}")
            return None
        
        async with rate_limiter:
            for attempt in range(max_attempts):
                try:
                    timeout = aiohttp.ClientTimeout(total=10 * (attempt + 1))
                    async with self.session.get(url, params=params, headers=headers, timeout=timeout) as response:
                        if response.status == 429:
                            wait_time = backoff_factor ** attempt
                            logger.warning(f"Rate limit exceeded for {provider_name}, retrying in {wait_time}s...")
                            await asyncio.sleep(wait_time)
                            continue
                        response.raise_for_status()
                        return await response.json()

                except aiohttp.ClientResponseError as e:
                    if attempt == max_attempts - 1:
                        logger.error(f"Request failed after {max_attempts} attempts: {e}")
                        raise
                    wait_time = backoff_factor ** attempt
                    logger.warning(f"Request attempt {attempt + 1} failed: {e}. Retrying in {wait_time}s...")
                    await asyncio.sleep(wait_time)
                except aiohttp.ClientConnectionError as e:
                    if attempt == max_attempts - 1:
                        logger.error(f"Connection error after {max_attempts} attempts: {e}")
                        raise
                    wait_time = backoff_factor ** attempt
                    logger.warning(f"Connection error on attempt {attempt + 1}: {e}. Retrying in {wait_time}s...")
                    await asyncio.sleep(wait_time)
                except Exception as e:
                    logger.error(f"Unexpected error: {e}")
                    raise

    async def fetch_historical_prices(self, token: str, days: int = 30) -> List[float]:
        """Fetch historical price data for a given token symbol."""
        cache_key = f"historical_prices_{token}_{days}"
        if cache_key in self.price_cache:
            logger.debug(f"Returning cached historical prices for {token}.")
            return self.price_cache[cache_key]
        prices = await self._fetch_from_services(
            lambda service: self._fetch_historical_prices(service, token, days),
            f"historical prices for {token}",
        )
        if prices:
            self.price_cache[cache_key] = prices
        return prices or []

    async def _fetch_historical_prices(self, source: str, token: str, days: int) -> Optional[List[float]]:
        """Fetch historical prices from a specified source."""
        config = self.api_configs.get(source)
        if not config:
            logger.error(f"API source {source} not configured.")
            return None

        try:
            async with self.session.get(config["base_url"] + f"/coins/{token}/market_chart?vs_currency=eth&days={days}") as response:
                if response.status == 200:
                    data = await response.json()
                    prices = [price[1] for price in data.get("prices", [])]
                    logger.debug(f"Fetched historical prices from {source}: {prices}")
                    return prices
                else:
                    logger.error(f"Failed to fetch historical prices from {source}: Status {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Exception fetching historical prices from {source}: {e}")
            return None

    async def get_token_volume(self, token: str) -> float:
        """Get the 24-hour trading volume for a given token symbol."""
        cache_key = f"token_volume_{token}"
        if cache_key in self.price_cache:
            logger.debug(f"Returning cached trading volume for {token}.")
            return self.price_cache[cache_key]
        volume = await self._fetch_from_services(
            lambda service: self._fetch_token_volume(service, token),
            f"trading volume for {token}",
        )
        if volume is not None:
            self.price_cache[cache_key] = volume
        return volume or 0.0

    async def _fetch_token_volume(self, source: str, token: str) -> Optional[float]:
        """Fetch token volume from a specified source."""
        config = self.api_configs.get(source)
        if not config:
            logger.debug(f"API configuration for {source} not found.")
            return None
        if source == "coingecko":
            url = f"{config['base_url']}/coins/markets"
            params = {"vs_currency": "usd", "ids": token}
            response = await self.make_request(source, url, params=params)
            return response[0]["total_volume"] if response else None
        else:
            logger.debug(f"Unsupported volume source: {source}")
            return None

    async def _fetch_from_services(self, fetch_func, description: str):
        """Helper method to fetch data from multiple services."""
        for service in self.api_configs.keys():
            try:
                logger.debug(f"Fetching {description} using {service}...")
                result = await fetch_func(service)
                if result:
                    return result
            except Exception as e:
                logger.warning(f"Failed to fetch {description} using {service}: {e}")
        logger.warning(f"Failed to fetch {description}.")
        return None

    async def _load_abi(self, abi_path: str) -> List[Dict[str, Any]]:
        """Load contract abi from a file."""
        try:
            async with aiofiles.open(abi_path, 'r') as file:
                content = await file.read()
                abi = json.loads(content)
            logger.debug(f"Loaded abi from {abi_path} successfully.")
            return abi
        except Exception as e:
            logger.error(f"Failed to load abi from {abi_path}: {e}")
            raise

    async def close(self):
        """Close the aiohttp session."""
        await self.session.close()

    async def initialize(self) -> None:
        """Initialize API configuration."""
        try:
            self.session = aiohttp.ClientSession()
            logger.info("API_Config initialized ✅")
        except Exception as e:
            logger.critical(f"API_Config initialization failed: {e}")
            raise

#//////////////////////////////////////////////////////////////////////////////